pkg/services/ngalert/notifier/channels/wecom.go

```diff
@@ -90,35 +135,119 @@ func (w *WeComNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, e
 	tmpl, _ := TmplText(ctx, w.tmpl, as, w.log, &tmplErr)
 
 	bodyMsg := map[string]interface{}{
-		"msgtype": "markdown",
+		"msgtype": w.settings.MsgType,
 	}
 	content := fmt.Sprintf("# %s\n%s\n",
 		tmpl(w.settings.Title),
 		tmpl(w.settings.Message),
 	)
+	if w.settings.MsgType != defaultWeComMsgType {
+		content = fmt.Sprintf("%s\n%s\n",
+			tmpl(w.settings.Title),
+			tmpl(w.settings.Message),
+		)
+	}
 
-	bodyMsg["markdown"] = map[string]interface{}{
+	bodyMsg[w.settings.MsgType] = map[string]interface{}{
 		"content": content,
 	}
 
+	url := w.settings.URL
+	if w.settings.channel != defaultWeComChannelType {
+		bodyMsg["agentid"] = w.settings.AgentID
+		bodyMsg["touser"] = w.settings.ToUser
+		token, err := w.GetAccessToken()
+		if err != nil {
+			return false, err
+		}
+		url = fmt.Sprintf(w.settings.EndpointURL+"/cgi-bin/message/send?access_token=%s", token)
+	}
+
 	body, err := json.Marshal(bodyMsg)
 	if err != nil {
 		return false, err
 	}
 
+	if tmplErr != nil {
+		w.log.Warn("failed to template WeCom message", "err", tmplErr.Error())
+	}
+
 	cmd := &models.SendWebhookSync{
-		Url:  w.settings.URL,
+		Url:  url,
 		Body: string(body),
 	}
 
-	if err := w.ns.SendWebhookSync(ctx, cmd); err != nil {
+	if err = w.ns.SendWebhookSync(ctx, cmd); err != nil {
 		w.log.Error("failed to send WeCom webhook", "err", err, "notification", w.Name)
 		return false, err
 	}
 
 	return true, nil
 }
 
+// GetAccessToken returns the access token for apiapp
+func (w *WeComNotifier) GetAccessToken() (string, error) {
```

Comment:
Please add the `ctx` parameter to this function and pass it through to the single-flighted getAccessToken call. Then we have cancellation control all the way down.